### ISSUE TITLE
XML and tabular output: off by one

### ISSUE THREAD

**Issue by lucventurini** [bbuchfink/diamond#334]
I am noticing that the DIAMOND XML and tabular output have different query and target HSP ends,
at least in BLASTX.

Tabular:
`cuff_cufflinks_star_at.23555.2	Q9FJX9	93.6	598	0	1	88	1881	1	560	0.0e+00	1120.1`

XML:
```xml
  <Iteration_iter-num>15</Iteration_iter-num>
  <Iteration_query-ID>Query_15</Iteration_query-ID>
  <Iteration_query-def>cuff_cufflinks_star_at.23555.2</Iteration_query-def>
  <Iteration_query-len>1902</Iteration_query-len>
<Iteration_hits>
<Hit>
  <Hit_num>1</Hit_num>
  <Hit_id>Q9FJX9</Hit_id>
  <Hit_def>RecName: Full=Protein DA1-related 7;</Hit_def>
  <Hit_accession>Q9FJX9</Hit_accession>
  <Hit_len>560</Hit_len>
  <Hit_hsps>
    <Hsp>
      <Hsp_num>1</Hsp_num>
      <Hsp_bit-score>1120.1</Hsp_bit-score>
      <Hsp_score>2896</Hsp_score>
      <Hsp_evalue>0.0e+00</Hsp_evalue>
      <Hsp_query-from>88</Hsp_query-from>
      <Hsp_query-to>1882</Hsp_query-to>
      <Hsp_hit-from>1</Hsp_hit-from>
      <Hsp_hit-to>560</Hsp_hit-to>
      <Hsp_query-frame>1</Hsp_query-frame>
      <Hsp_hit-frame>0</Hsp_hit-frame>
      <Hsp_identity>560</Hsp_identity>
      <Hsp_positive>560</Hsp_positive>
      <Hsp_gaps>38</Hsp_gaps>
      <Hsp_align-len>598</Hsp_align-len>
          </Hsp>
```

Tabular reports 1881, whereas the XML reports 1882.

In the NCBI BLAST+, the reported end is the same in both formats.

**Comment by lucventurini** [comment]
Update: upon further testing, the rule seems to be that when the hit is on the positive strand (ie qend > qstart), then qend is off by 1. When the hit is on the negative strand (ie qstart > qend) then *qstart* is off by 1.

I am patching in my workflows using `awk`, for the time being.

**Comment by bbuchfink** [comment]
Thanks for reporting this, it should be fixed in adc14a30631955e17055aac4ec96814187b22d5f.

For me it was always qend that is off by 1. I also checked it against Blast. Let me know if this works for you.

By the way, for automated parsing I would definitely recommend using the tabular format. You should be able to get all the info you need from that format.

**Comment by lucventurini** [comment]
Thank you for fixing this. We are definitely moving towards using always tabular format in automated workflows, but we'll leave the XML input option available.
We are already seeing a significant boost in performance, which is great.
